### PR TITLE
ambient: ensure we sync network before we mark ourselves synced

### PR DESCRIPTION
**Pull request by howardjohn**
otherwise, we can startup without network gateway sync and give
incorrect response

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go [hunk 1]
```diff
@@ -173,9 +173,9 @@ func (a *index) workloadEntryWorkloadBuilder(
 
 		if addr, err := netip.ParseAddr(wle.Spec.Address); err == nil {
 			w.Addresses = [][]byte{addr.AsSlice()}
-		} else {
+		} else if wle.Spec.Address != "" {
 			log.Warnf("skipping workload entry %s/%s; DNS Address resolution is not yet implemented", wle.Namespace, wle.Name)
-		}
+		} // Else it is an empty address with network set, this is ok
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:178]: with network gateway

File: pkg/kube/krt/recomputetrigger.go [hunk 2]
```diff
@@ -50,3 +50,10 @@ func (r *RecomputeTrigger) TriggerRecomputation() {
 func (r *RecomputeTrigger) MarkDependant(ctx HandlerContext) {
 	_ = Fetch(ctx, r.inner.AsCollection())
 }
+
+// MarkSynced marks this trigger as ready. Before this is called, dependant collections will be blocked.
+// This ensures initial state is populated.
+func (r *RecomputeTrigger) MarkSynced() *RecomputeTrigger {
```
Comment (hzxuzhonghu) [inline pkg/kube/krt/recomputetrigger.go:56]: Return val is not needed

**Review by hzxuzhonghu — COMMENTED** [review]
Is there any integration test now

**Comment by howardjohn** [comment]
currently only the control plane is implemented not ztunnel, so there
cannot be integration tests yet

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 3]
```diff
@@ -649,6 +652,18 @@ func (c *Controller) Run(stop <-chan struct{}) {
 	go c.exports.Run(stop)
 	kubelib.WaitForCacheSync("kube controller", stop, c.informersSynced)
 	log.Infof("kube controller for %s synced after %v", c.opts.ClusterID, time.Since(st))
+	if c.ambientIndex != nil {
```
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:655]: Is there a way to describe this behavior with a unit test in `controller_test.go`?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:655]: Added.. its not pretty but maybe good enough

File: pkg/kube/krt/recomputetrigger.go (hunk 2 above)
Comment (bleggett) [inline pkg/kube/krt/recomputetrigger.go:56]:
```suggestion
func (r *RecomputeTrigger) MarkSynced() {
```

(to echo @hzxuzhonghu - AFAICT this is not needed)

File: pkg/kube/krt/recomputetrigger.go [hunk 4]
```diff
@@ -50,3 +50,10 @@ func (r *RecomputeTrigger) TriggerRecomputation() {
...
+	r.inner.MarkSynced()
+	return r
```
Comment (bleggett) [inline pkg/kube/krt/recomputetrigger.go:58]:
```suggestion
```

**Review by bleggett — APPROVED** [review]

**Review by stevenctl — APPROVED** [review]

**Pull request merged**